sycl/doc/extensions/C-CXX-StandardLibrary/C-CXX-StandardLibrary.rst

```diff
@@ -0,0 +1,162 @@
+
+  - in Glibc and musl libraries it expands to `__assert_fail`
+  - in MSVC library it expands to `_wassert`
+  - in newlib library it expands to `__assert_func`
+
+This makes it difficult to handle all possible cases in device
+compilers. In order to facilitate porting to new platforms, and to
+avoid imposing a lot of boilerplate code in *every* device compiler,
+wrapper libraries are provided with the SYCL compiler that "lower"
+libc implementation-specific functions into a stable set of functions,
+that can be later handled by a device compiler.
+
+.. code:
+   clang++ -fsycl -c main.cpp -o main.o
+   clang++ -fsycl main.o $(SYCL_INSTALL)/lib/libsycl-glibc.o -o a.out
+
+This `libsycl-glibc.o` is one of these wrapper libraries: it provides
+definitions for glibc specific library function, and these definitions
+call the corresponding functions from `__devicelib_*` set of
+functions.
+
+For example, `__assert_fail` from IR above gets transformed into:
+.. code:
+    ; Function Attrs: noreturn nounwind
+    declare dso_local spir_func void @__devicelib_assert_fail(i8 addrspace(4)*, i8 addrspace(4)*, i32, i8 addrspace(4)*)
+
+    ; Function Attrs: noreturn nounwind
+    define dso_local spir_func void @__assert_fail(i8 addrspace(4)*, i8 addrspace(4)*, i32, i8 addrspace(4)*) {
+      call spir_func void @__devicelib_assert_fail([...])
+    }
+
+    [...]
+    cond.false:
+      call spir_func void @__assert_fail([...])
+      unreachable
+
+A single wrapper object provides function wrappers for *all* supported
+library functions. Every supported C library implementation (MSVC or
+glibc) has its own wrapper library object:
+
+  - libsycl-glibc.o
+  - libsycl-msvc.o
+
+SPIR-V
+======
+
+Standard library functions are represented as external (import)
+functions in SPIR-V:
+
+.. code:
+   8 Decorate 67 LinkageAttributes "__devicelib_assert_fail" Import
+   ...
+   2 Label 846
+   8 FunctionCall 63 864 67 855 857 863 859
+   1 Unreachable
+
+Device compiler
+===============
+
+Device compiler is free to implement these `__devicelib_*` functions.
+In order to indicate support for a particular set of functions,
+underlying runtime have to support the corresponding OpenCL (PI)
+extension. See ``DeviceLibExtensions.rst`` for a list of supported
+functions and corresponding extensions.
+
+Fallback implementation
+=======================
+
+If a device compiler does not indicate "native" support for a
+particular function, a fallback library is linked at JIT time by the
+SYCL Runtime. This library is distributed with the SYCL Runtime and
+resides in the same directory as the `libsycl.so` or `sycl.dll`.
+
+A fallback library is implemented as a device-agnostic SPIR-V program,
+and it is supposed to work for any device that supports SPIR-V.
+
+Every set of functions is implemented in a separate fallback
+library. For example, a fallback for `cl_intel_devicelib_cassert`
+extension is provided as `libsycl-fallback-cassert.spv`
```

Comment:
Do we really need to move "Every set of functions" to a separate library? Can't it be handled by the linker?
Will it work if we pass something like `-l<native lib> -l<fallback lib>` and linker looks for function implementation in native lib first and if it's not found, query fallback library?

Reply:
There is no <native lib> in SYCL compiler - a device compiler provides one, similarly to how OpenCL built-in functions are handled.